File: python_app/services/nominatim.py
import requests

NOMINATIM_URL = "https://nominatim.openstreetmap.org/search"

def geocode_place(query):
    params = {"format": "jsonv2", "q": query, "limit": 1, "addressdetails": 0}
    headers = {"Accept": "application/json", "User-Agent": "tourism-multi-agent-python/1.0"}
    try:
        r = requests.get(NOMINATIM_URL, params=params, headers=headers, timeout=20)
    except Exception:
        return None
    if r.status_code != 200:
        return None
    try:
        arr = r.json()
    except Exception:
        return None
    if not isinstance(arr, list) or not arr:
        return None
    x = arr[0]
    lat = x.get("lat")
    lon = x.get("lon")
    if not lat or not lon:
        return None
    bbox_arr = x.get("boundingbox", [])
    bbox = None
    if isinstance(bbox_arr, list) and len(bbox_arr) == 4:
        bbox = {
            "south": float(bbox_arr[0]),
            "north": float(bbox_arr[1]),
            "west": float(bbox_arr[2]),
            "east": float(bbox_arr[3]),
        }
    return {"name": x.get("display_name", query), "lat": float(lat), "lon": float(lon), "bbox": bbox}

File: src/services/nominatim.js
const NOMINATIM_URL = "https://nominatim.openstreetmap.org/search";

export async function geocodePlace(query) {
  const u = new URL(NOMINATIM_URL);
  u.searchParams.set("format", "jsonv2");
  u.searchParams.set("q", query);
  u.searchParams.set("limit", "1");
  u.searchParams.set("addressdetails", "0");
  const res = await fetch(u.toString(), {
    headers: {
      "Accept": "application/json",
      "User-Agent": "tourism-multi-agent/1.0"
    }
  });
  if (!res.ok) return null;
  const arr = await res.json();
  if (!Array.isArray(arr) || arr.length === 0) return null;
  const r = arr[0];
  if (!r || !r.lat || !r.lon) return null;
  const bboxArr = r.boundingbox || [];
  const bbox = bboxArr.length === 4 ? {
    south: parseFloat(bboxArr[0]),
    north: parseFloat(bboxArr[1]),
    west: parseFloat(bboxArr[2]),
    east: parseFloat(bboxArr[3])
  } : null;
  return {
    name: r.display_name || query,
    lat: parseFloat(r.lat),
    lon: parseFloat(r.lon),
    bbox
  };
}